### PR TITLE
Change reader.Read in readObjectFile to fix #18

### LINKED ISSUE #18: Buffer read change in Go 1.7 breaks readObjectFile()
I found that after upgrading to Go 1.7, `repo.GetCommitOfBranch(branch)` started returning an `EOF` and not the commit sha1 that I expected. 

https://github.com/gogits/git/blob/564c5fe86bbef8589421582b69657cc08c4c85fb/repo_utils.go#L306

I dug in and found that simply removing the error check fixes this, which matches up with the release notes below. 

> In previous releases of Go, if Reader's Read method were asked for zero bytes with no data remaining, it would return a count of 0 and no error. Now it returns a count of 0 and the error io.EOF

### PR DESCRIPTION
**Pull request by snicol**
See https://github.com/gogits/git/issues/18

### CODE REVIEW

File: repo_utils.go [hunk 1]
```diff
@@ -303,8 +303,8 @@ func readObjectFile(path string, sizeonly bool) (ot ObjectType, length int64, da
 	firstBufferSize := int64(1024)
 
 	buf := make([]byte, firstBufferSize)
-	_, err = r.Read(buf)
-	if err != nil {
+	nBytes, err := r.Read(buf)
```
Comment (Unknwon) [inline repo_utils.go]: Skipping the error does not seem a good practice?

**Comment by snicol** [comment]
As it will return an io.EOF when done the error needs multiple checks right? Something like

    if err != nil || err != io.EOF

**Comment by Unknwon** [comment]
Yes, I think we should first:

1. Read any bytes left.
2. Check err, return if not io.EOF
3. Break out if it's io.EOF.

Sounds right?...

**Comment by snicol** [comment]
Sounds good, I'm AFK until tomorrow so will change then 😊

**Comment by lemmi** [comment]
what about using something like https://golang.org/pkg/bufio/#Reader.ReadBytes and delimit on a space. that way a lot of work can be offloaded to the stdlib and it's probably less error prone.

**Comment by lemmi** [comment]
Ok, hang on. I had another look at the function and I'm not sure wether it's leaking the file. the way I see it:

1. file opened here: https://github.com/gogits/git/blob/master/repo_utils.go#L285
2. if there's an error, gets closed https://github.com/gogits/git/blob/master/repo_utils.go#L290-L296
3. gets wrapped into a zlib reader https://github.com/gogits/git/blob/master/repo_utils.go#L298
4. get reseeked to the beginning https://github.com/gogits/git/blob/master/repo_utils.go#L335
5. gets wrapped into a zlib reader again https://github.com/gogits/git/blob/master/repo_utils.go#L340

From this point on there is no direct access to f anymore, only behind io.Readers. I already started on fixing this and then saw the same pattern in https://github.com/gogits/git/blob/master/repo_utils.go#L172

So i propose another solution. 

1. I'd like to change readerDecompressed to take an io.ReadCloser and make the Close() method to close both the decompressor and the underlying file.
2. Rewrite the function and get rid of the seeking and using 2 zlib decompressors.

The changes aren't done yet. I'll make another pull request as soon as I can.

**Comment by Unknwon** [comment]
@lemmi good catch!

**Pull request closed without merging**